libs/core/src/lib/content-density/classes/content-density-consumer.class.ts

```diff
@@ -0,0 +1,140 @@
+
+/** Class, which is used for consuming calculated content density state information in components */
+export class ContentDensityObserver extends BehaviorSubject<ContentDensityMode> {
+    /** Stream for getting compact state */
+    isCompact$: Observable<boolean>;
+    /** Stream for getting cozy state */
+    isCozy$: Observable<boolean>;
+    /** Stream for getting condensed state */
+    isCondensed$: Observable<boolean>;
+
+    callbacks: Map<any, ContentDensityCallbackFn> = new Map<any, ContentDensityCallbackFn>();
+
+    /** @hidden */
+    private _isCompact: boolean;
+    /** @hidden */
+    private _isCozy: boolean;
+    /** @hidden */
+    private _isCondensed: boolean;
+
+    /** @hidden */
+    private _compactSubscription: Subscription;
+    /** @hidden */
+    private _condensedSubscription: Subscription;
+    /** @hidden */
+    private _cozySubscription: Subscription;
+
+    /** @hidden */
+    private _callCallbacks$ = new BehaviorSubject<void>(undefined);
+
+    /** @hidden */
+    constructor(
+        initialValue: ContentDensityMode,
+        dataSource: Observable<ContentDensityMode>,
+        private destroy$: Observable<void>,
+        private changeDetectorRef: ChangeDetectorRef
+    ) {
+        super(initialValue);
+        const source$ = dataSource.pipe(distinctUntilChanged(), takeUntil(this.destroy$));
+        combineLatest([source$, this._callCallbacks$])
+            .pipe(
+                tap(([density]) => this._callCallbacks(density)),
+                map(([density]) => density),
+                finalize(() => this.complete())
+            )
+            .subscribe((density) => {
+                this.next(density);
+            });
+        this.isCompact$ = source$.pipe(map(isCompact));
+        this.isCozy$ = source$.pipe(map(isCozy));
+        this.isCondensed$ = source$.pipe(map(isCondensed));
+    }
+
+    /**
+     * Sync isCompact. Calls ChangeDetectorRef.markForCheck() on update
+     */
+    get isCompact(): boolean {
+        if (!this._compactSubscription) {
+            this._compactSubscription = this._listenToChanges(this.isCompact$, (v) => (this._isCompact = v));
+        }
+        return this._isCompact;
+    }
+
+    /**
+     * Sync isCondensed. Calls ChangeDetectorRef.markForCheck() on update
+     */
+    get isCondensed(): boolean {
+        if (!this._condensedSubscription) {
+            this._condensedSubscription = this._listenToChanges(this.isCondensed$, (v) => (this._isCondensed = v));
+        }
+        return this._isCondensed;
+    }
+
+    /**
+     * Sync isCondensed. Calls ChangeDetectorRef.markForCheck() on update
+     */
+    get isCozy(): boolean {
+        if (!this._cozySubscription) {
+            this._cozySubscription = this._listenToChanges(this.isCozy$, (v) => (this._isCozy = v));
+        }
+        return this._isCozy;
+    }
+
+    /**
+     * Pass consumers callback function, or configuration object to be called on density change
+     */
+    consume(
```

Comment:
i cannot fully understand why do we have this. 

example. we may have component `A` that is a **consumer**. classes applied and so on. 

_first scenario_: 
if there is a nested comp `B` - we can make it consumer too. 

_second scenario_: 
if there are some elements in template of `A` (not a components for some reason) and we want to apply classes to them depending on content density we could write `[class.cozy]="_contentDensityConsumer.isCozy$ | async"` and that's it. 
if we want to run custom code we may do `_contentDensityConsumer.isCozy$.subscribe(() => { ... })` and that's it.
or we could create a template variable, query for it with `ContentChild()` and only then pass it to the `_contentDensityConsumer.consume(...)` and we will have the same result?

Reply:
Yeah, I encountered such scenarios, where content density classes are added to elements, which did not have directive or component, that's why If someone wants to have everything focused in TS, I created that support. If they want to have it on HTML, that is totally possible too

Reply:
Plus, it's open source, I do not know what weird things are other people doing, they may be creating elements through document,createElement and then injecting in DOM, so if they want to add content density to their components, they can include that